### ISSUE TITLE
Auth.updateUserAttributes not working

### ISSUE THREAD

**Issue by lindachenau** [aws-amplify/amplify-js#6329]
**Describe the bug**
Auth.updateUserAttributes return Unhandled Rejection (TypeError): user.getSession is not a function

**To Reproduce**
Execute the code snippet with the following packages always produced the same error.

"aws-amplify": "^3.0.13",
"aws-amplify-react": "^4.1.12",
"react": "^16.6.3",

**Code Snippet**
let user = await Auth.currentAuthenticatedUser()
const session = await Auth.userSession(user)
user.Session = session
await Auth.updateUserAttributes({
  user, 
  attributes: { email, phone_number },
}))

**Additional context**
I've read the discussions in #2276 and #238. The CognitoUser object returned by Auth.currentAuthenticatedUser() still has Session object set to null. That's why I tried manually setting Session object to the CognitoUserSession object returned by Auth.userSession(user). It doesn't change the result. I'm still getting user.getSession is not a function error.

The  CognitoUser object is shown below. CognitoUserSession object came from Auth.userSession(user). Otherwise it is null.

username: "98659cc9-9e29-4aed-b4f4-ea745cff0feb"
pool: CognitoUserPool {userPoolId: "ap-southeast-2_vSlGDgdH9", clientId: "6emoisvvj2gnmhiauoqgneirnq", client: Client, advancedSecurityDataCollectionFlag: true, storage: Storage}
Session: CognitoUserSession {idToken: CognitoIdToken, refreshToken: CognitoRefreshToken, accessToken: CognitoAccessToken, clockDrift: 0}
client: Client {endpoint: "https://cognito-idp.ap-southeast-2.amazonaws.com/", fetchOptions: {…}}
signInUserSession: CognitoUserSession {idToken: CognitoIdToken, refreshToken: CognitoRefreshToken, accessToken: CognitoAccessToken, clockDrift: 0}
authenticationFlowType: "USER_SRP_AUTH"
storage: Storage {CognitoIdentityServiceProvider.4kob0fpip2l7afbdups4m0scs6.767d1df3-c0e7-4ef0-be56-6b120918391c.userData: "{"UserAttributes":[{"Name":"sub","Value":"767d1df3…Username":"767d1df3-c0e7-4ef0-be56-6b120918391c"}", CognitoIdentityServiceProvider.4kob0fpip2l7afbdups4m0scs6.724ed80b-5c19-4300-b4a6-e2ca08a345a4.clockDrift: "1", CognitoIdentityServiceProvider.6emoisvvj2gnmhiauoqgneirnq.LastAuthUser: "98659cc9-9e29-4aed-b4f4-ea745cff0feb", aws.cognito.identity-id.ap-southeast-2:f84bb7b3-17fb-47bb-bf79-c311ceae5386: "ap-southeast-2:7b8d3b7c-e13c-47ba-9f65-960e4b26eaa3", gatsbyUser: "{"sub":"98659cc9-9e29-4aed-b4f4-ea745cff0feb","ema…username":"98659cc9-9e29-4aed-b4f4-ea745cff0feb"}", …}
keyPrefix: "xxxx"
userDataKey: "xxxx"
attributes: {sub: "98659cc9-9e29-4aed-b4f4-ea745cff0feb", email_verified: true, phone_number_verified: false, email: "lindachenau@gmail.com"}
preferredMFA: "NOMFA"

"aws-amplify": "^3.0.13",
"aws-amplify-react": "^4.1.12",
"react": "^16.6.3",

**Comment by giulio-dds** [comment]
Hi, I encountered the same problem and I solved it, inspired by your code. I don't think this solution can be forever stable but for now it could be a good help.

```
const userFromCognito = await Auth.currentAuthenticatedUser()
userFromCognito.Session = userFromCognito.signInUserSession
await Auth.updateUserAttributes(userFromCognito, { "custom:attribute": "value" })
```

**Comment by lindachenau** [comment]
@giulio-dds Hi, thanks for sharing your solution. Unfortunately I still got the same error after trying out your solution. I hope someone from the amplify team can provide a solution.

**Comment by Rouche01** [comment]
Please, I need a working solution for this issue. I have tried all the methods I have seen here and on the documentation and none of them seem to work

**Comment by lindachenau** [comment]
@Rouche01 I was hoping someone from the amplify team can offer a solution. Unfortunately no one answer it.

**Comment by Rouche01** [comment]
Hopefully, they will offer a solution soon

**Comment by Rouche01** [comment]
@lindachenau I found a solution to work around this problem using the 'amazon-cognito-identity-js' npm package.

`import { CognitoUserAttribute, CognitoUserPool } from 'amazon-cognito-identity-js';
const poolData = {
  UserPoolId: 'pool_id',
  ClientId: 'client_id'
};

const userPool = new CognitoUserPool(poolData);
const attributeList = [];
const attribute = {
    Name: 'nickname',
    Value: 'joe',
};

const attribute = new CognitoUserAttribute(attribute);
attributeList.push(attribute);

cognitoUser.updateAttributes(attributeList, function(err, result) {
    if (err) {
        alert(err.message || JSON.stringify(err));
        return;
    }
    console.log('call result: ' + result);
});`

You can check out the npm usage documentation to understand its usage more clearly

**Comment by Rouche01** [comment]
@lindachenau I found a solution to work around this problem using the 'amazon-cognito-identity-js' npm package.

```
import { CognitoUserAttribute, CognitoUserPool } from 'amazon-cognito-identity-js';
const poolData = {
  UserPoolId: 'pool_id',
  ClientId: 'client_id'
};

const userPool = new CognitoUserPool(poolData);
const attributeList = [];
const attribute = {
    Name: 'nickname',
    Value: 'joe',
};

const attribute = new CognitoUserAttribute(attribute);
attributeList.push(attribute);

cognitoUser.updateAttributes(attributeList, function(err, result) {
    if (err) {
        alert(err.message || JSON.stringify(err));
        return;
    }
    console.log('call result: ' + result);
});
```

You can check out the npm usage documentation to understand its usage more clearly

**Comment by lindachenau** [comment]
@Rouche01 Thanks so much for this solution. I'll try it out.

**Comment by lindachenau** [comment]
@Rouche01 I figured it out that you don't have to replace authentication method. The following code works for me. Thank you so much for finding this solution. 

import {
  CognitoUserPool,
  CognitoUserAttribute
} from 'amazon-cognito-identity-js'

const poolData = {
  UserPoolId: 'pool_id',
  ClientId: 'client_id', 
}

const userPool = new CognitoUserPool(poolData)

const authenticatedUser = await userPool.getCurrentUser()

if (authenticatedUser != null) {
  authenticatedUser.getSession(function(err, session) {
    if (err) {
      alert(err)
      return
    }
    console.log('session validity: ' + session.isValid())
    
    let attributeList = []

    const attribute = new CognitoUserAttribute({
      Name: 'phone_number',
      Value: '12345678',
    })
    
    attributeList.push(attribute)   

    authenticatedUser.updateAttributes(attributeList, function(err, result) {
      if (err) {
        alert(err.message || JSON.stringify(err))
        return
      }
      console.log('call result: ' + result)
    })
  })
}

**Comment by ta1bbty** [comment]
Hi @lindachenau,
I tried this doesn't seem to work. 
Also, how are you getting the current user in the above code :thinking: 
```js
const authenticatedUser = await userPool.getCurrentUser()
```
I replaced this with `Auth.currentAuthenticatedUser` this also seems to return `CognitoUser`
However, i doubt if both the classes ( `CognitoUser` ) are same or not.

**Comment by lindachenau** [comment]
I believe they are the same because my user is signed in with Auth.signin before I call userPool.getCurrentUser()

**Comment by Salmandabbakuti** [comment]
I'm wondering why we're using ```Auth.currentAuthenticatedUser()``` to update user attributes? what if there's a new user ```user2``` who just logs in while we're in process of updating user ```user1``` attributes. so the ```Auth.currentAuthenticatedUser()``` will now points to the user ```user2``` who just logs in and we end up updating new user attributes(```user2```) instead of ours? (```user1```).  isn't that a big issue?
@sammartinez @amhinson @elorzafe

**Comment by jackoconnor21** [comment]
@Salmandabbakuti Auth.currentAuthenticatedUser() will only return the current authenticated user on the machine it's called from. Even if you logged out of that user and logged in as another within the time of making that call (damn near impossible to do), the response would already be stored. What scenario are you thinking that could happen in?

**Comment by Salmandabbakuti** [comment]
@jackoconnor21 I don't think that Auth.currentAuthenticatedUser() will only return the current authenticated user on the machine it's called from instead it is specific to the pool. correct me if i'm wrong

**Comment by dylan-authentic** [comment]
@sammartinez @elorzafe Any updates on a working solution for the original issue? `Auth.updateUserAttributes(user, { "custom:attribute": "value" })` is still returning `user.getSession is not a function`

**Comment by Niretech** [comment]
@Salmandabbakuti - after logging in to multiple users on the same machine, with different browsers (incognito), I can confirm that Auth.currentAuthenticatedUser() only returns the user related to the current session.

@dylan-authentic To run updateUserAttributes you don't need to associate a session with the user object.. unless you're doing so for a different reason.
